### PR TITLE
fix(sqlite): Fix removeColumn and changeColumn foreign key issues for the sqlite dialect

### PR DESCRIPTION
**Pull request by martineso**
### Pull Request Checklist

_Please make sure to review and check all of these items:_

- [x] Have you added new tests to prevent regressions?
- [x] Does `yarn test` or `yarn test-DIALECT` pass with this change (including linting)?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [?] Did you update the typescript typings accordingly (if applicable)?
  There is missing type definition for the return type of `queryInterface.tableDescription`. I have not updated it though since it was missing other type information before I have made any changes and I did not want to mess up any tests.
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

### Description Of Change

This PR addresses the following issue #12741. It also includes a fix for `queryInterface.changeColumn` of the **sqlite** dialect.

### Todos

- [ ] Maybe add TS definitions for `queryInterface.tableDescription` for **sqlite** only?

### CODE REVIEW

**Comment by martineso** [comment]
Can someone take a look at the failing integration tests and if they are due to a change I have made? I have made sure to only touch the tests related to **sqlite** however I may have caused a problem with other dialects unknowingly and I need a second pair of eyes to check. Thanks!

**Comment by WikiRik** [comment]
They're failing because they were only intended for sqlite. If you move the check for sqlite up by two tests you should be in the green again (except for DB2 probably but we have very flaky tests in those)

**Comment by martineso** [comment]
I have resolved the issue with ```changeColumn.test.js``` for sqlite and I merged the latest **main**. 
There is however an issue with the ```transaction.test.js``` where for some reason the sqlite dialect fails for 
```
1) [SQLITE] Transaction
       fails with SQLITE_BUSY when retry.match is changed:
```
I noticed this test is marked as flaky in this issue #14161.
I am certain that the changes I've made are not related to this issue since the test is failing in **main** branch as well.

**Comment by WikiRik** [comment]
Yes, that failing test is not due to your PR. We'll review this soon

File: src/dialects/sqlite/query-interface.js
Comment (ephys) [inline src/dialects/sqlite/query-interface.js:26]: we should prefix our internal savepoints to be somewhat safe.
Or use a uuid, or a prefixed uuid. e.g. `sequelize_3958bbe4-284a-4d2f-9c5c-b136adc95774`

File: src/dialects/sqlite/query-interface.js
Comment (ephys) [inline src/dialects/sqlite/query-interface.js:31]: if this fails for any reason (such as network issue). `PRAGMA foreign_keys` will not be reverted.

There are a few places where `PRAGMA foreign_keys` is used. It would be better extracted to a helper function:

```typescript
async withForeignKeysOff(options, cb) {
  await this.sequelize.query('PRAGMA foreign_keys = OFF', options);
  
  try {
    return await cb();
  } finally {
    await this.sequelize.query('PRAGMA foreign_keys = ON', options);
  }
}

// ...

this.withForeignKeysOff(options, () => {
  await this.sequelize.query(`SAVEPOINT ${savepointName};`, options);
  // ...
});
```

File: src/dialects/sqlite/query-interface.js
Comment (ephys) [inline src/dialects/sqlite/query-interface.js:38]: can these be executed in parallel?

File: src/dialects/sqlite/query-interface.js
Comment (ephys) [inline src/dialects/sqlite/query-interface.js:85]: A lot of code here is shared with `removeColumn`, can they be deduplicated?

**Review by ephys — CHANGES_REQUESTED** [review]
Looking good, I have a few notes :)

File: src/dialects/sqlite/query-interface.js
Comment (martineso) [reply src/dialects/sqlite/query-interface.js:38]: These should be executed sequentially since the order of execution matters.

File: src/dialects/sqlite/query-interface.js
Comment (martineso) [reply src/dialects/sqlite/query-interface.js:85]: I was thinking the same thing while writing the code and I had the following dilemma. 
Extracting the logic to a common function will reduce the overall code however it will increase the cognitive load due to the added abstraction. I may however rethink the way we can extract this logic to a separate function using your suggestion for the ```withForeignKeysOff``` function

File: src/dialects/sqlite/query-interface.js
Comment (martineso) [reply src/dialects/sqlite/query-interface.js:31]: Looks good, I will redo the logic inside ```changeColumn``` and ```removeColum``` using the above abstraction.

File: src/dialects/sqlite/query-interface.js
Comment (martineso) [reply src/dialects/sqlite/query-interface.js:26]: Right, makes sense. 
Is there a helper function in the repo for generating uuids?

Comment (ephys) [reply src/dialects/sqlite/query-interface.js:26]: the uuid package is available :)

**Comment by martineso** [comment]
Is there any progress on this PR?

File: test/integration/query-interface/changeColumn.test.js
Comment (ephys) [inline test/integration/query-interface/changeColumn.test.js:246]: I think we can run these tests on all dialects (and only disable the dialects that fail)

File: test/integration/query-interface/changeColumn.test.js
Comment (ephys) [inline test/integration/query-interface/changeColumn.test.js:409]: Why not this?

```suggestion
        const constraintsQuery = this.queryInterface.showConstraint('users');
```

File: test/integration/query-interface/changeColumn.test.js
Comment (ephys) [inline test/integration/query-interface/changeColumn.test.js:401]: Can you also (or instead) apply a change to the `level_id` column? (such as update its nullability and/or its name) to make sure it does not lose its reference?

File: test/integration/query-interface/removeColumn.test.js
Comment (ephys) [inline test/integration/query-interface/removeColumn.test.js:94]: we may as well run this on all dialects, to be sure.

**Review by ephys — CHANGES_REQUESTED** [review]
This PR got lost in limbo for so long because I lost its notification. I'm very sorry :(

Overall this is a really high quality PR, it's great work!
I have a few final questions but this other than that it's ready to go

File: test/integration/query-interface/changeColumn.test.js
Comment (martineso) [reply test/integration/query-interface/changeColumn.test.js:409]: I have tried the ```this.queryInterface.showConstraint('users');``` approach but it does not work for sqlite. It seems the problem lies within `src/dialects/abstract/query-interface.js#showConstraint` function. The `QueryTypes.SHOWCONSTRAINTS` type returns empty result for the sqlite dialect. This poses a problem however if we are to support all dialects within these new tests. 
I have problems with porting these test cases for MariaDB and MySQL (possibly others) due to using the workaround for sqlite (the one used in the test case before the suggested change). If you have a workaround or a suggestion I would gladly apply it.

File: test/integration/query-interface/changeColumn.test.js
Comment (martineso) [reply test/integration/query-interface/changeColumn.test.js:246]: I have problems with porting these test cases for other dialects. I have described the problem in another comment below (as a reply to another suggestion of yours). I am not well versed within the sequelize code base so I would love if you can provide a way to port the tests for other dialects.

File: test/integration/query-interface/changeColumn.test.js
Comment (martineso) [reply test/integration/query-interface/changeColumn.test.js:409]: Would you prefer (1) a different test case for each dialect or (2) a branching logic within the same test depending on the dialect type? 

1. 
```js

if (dialect === 'sqlite) {
 it('sqlite specific version of this test', ...)
} else if (dialect === 'mysql') {
 it('mysql version of the test', ...)
}
...
```
2.
```js
it('some generic test', async () => {
  if (dialect === 'sqlite') {
    expect().some.condition()
  } else {
    expect().some.other.condition()
  }
}

```

Comment (martineso) [reply test/integration/query-interface/changeColumn.test.js:409]: Also there is a 3rd solution, to fix the `showConstraint` functionality for sqlite dialect altough this would be out of the scope of this PR.

Comment (ephys) [reply test/integration/query-interface/changeColumn.test.js:409]: Definitely option 2. However I would also prefer to fix `showConstraint` first, maybe in a parallel PR we can merge first? (if you're willing to help with that too)

Comment (martineso) [reply test/integration/query-interface/changeColumn.test.js:409]: Yup, sure. I will open a new PR with a fix for `showConstraint`. Can you point me to a type definition for a valid return type for this function?

Comment (ephys) [reply test/integration/query-interface/changeColumn.test.js:409]: I ran this locally and I concur that fixing sqlite's showConstraint is going to be very complicated and I don't want to block this PR until that minor issue is fixed

Can you add the following comment above the call to `showConstraintsQuery`?

`// TODO: replace with queryInterface.showConstraint once it lists foreign keys properly for sqlite`

File: test/integration/query-interface/changeColumn.test.js
Comment (ephys) [inline test/integration/query-interface/changeColumn.test.js:438]: `level` should be created first, it is needed by `users`

**Review by ephys — CHANGES_REQUESTED** [review]

File: test/integration/query-interface/changeColumn.test.js
Comment (ephys) [reply test/integration/query-interface/changeColumn.test.js:246]: I'm also currently working on `changeColumn` to fix many more bugs, I'll take a look at making these tests run in all dialects in my PR because I first need to fix a bunch more bugs in `showConstraint`, `describeTable` etc

Can you add a `// TODO: run in all dialects` above this condition?

File: test/integration/query-interface/changeColumn.test.js
Comment (martineso) [reply test/integration/query-interface/changeColumn.test.js:451]: Done!

File: test/integration/query-interface/changeColumn.test.js
Comment (martineso) [reply test/integration/query-interface/changeColumn.test.js:402]: Added a change for level_id -> changed the nullability.

File: test/integration/query-interface/changeColumn.test.js
Comment (martineso) [reply test/integration/query-interface/changeColumn.test.js:422]: Alright. I have added a comment as per your suggestion.

**Review by ephys — APPROVED** [review]
Good work, thank you for this :)

**Pull request merged**
